### PR TITLE
[Tooling] Enable pod push on BuildKite

### PR DESCRIPTION
**Pull request by AliSoftware**
Request ref: paaHJt-2Gr-p2

### What

Enables pod lint build & test, validate, lint, and pod push.

### Checklist

 - [x] Normalize podspec file
 - [X] Create the pipeline and script files in `.buildkite`
   - [x] Disable CDN push while we refine this PR, pushing on our internal spec repo instead
   - [x] Temporarily enable "Publish Podspec" step on this branch (and not just tag) while testing the PR
   - [x] Re-enable correct filter on "Publish Podspec" step to only be on tags
   - [x] Re-enable CDN push once everything is ready
 - [x] Configure BuildKite
   - [x] Create a pipeline in BuildKite with `pipeline-upload` step
   - [x] Configure the WebHooks
   - [x] Declare proper environment variables in mobile-secrets and upload them
   - [x] Create a sha-256 keypair, upload the pair to mobile-secrets, and the public key as Deploy key to GitHub

### CODE REVIEW

File: .buildkite/pipeline.yml
Comment (AliSoftware) [inline .buildkite/pipeline.yml]: Only while testing the PR. Remember to revert this before merge.

File: .buildkite/publish-pod.sh
Comment (AliSoftware) [inline .buildkite/publish-pod.sh]: Only disable CDN push while testing the PR. Remember to revert this before merge.

**Comment by AliSoftware** [comment]
### ✅ Tested in the following state:

This is what I did to be able to validate those changes – before cleaning things up and making the PR ready for review:

 - I included in the PR debug commits adf8ca8 (to disable `pod trunk push` and only push to our internal spec repo) and 00b6a9f (to switch the `podspec` to `s.version = '0.0.2'`). I also included 9a48b7a (to allow the pod push to run on this branch while working on it), but that ended up not being necessary because I needed to create a tag to test things anyway (see next point).
 - I created a temporary `0.0.2` tag on top of my branch, to trigger a build in Buildkite and make it pass validation (as `pod repo push` calls `pod lib lint` which ensures we point to a tag, amongst other things) and be published internally
 - It succeeded and posted the notification in our Slack (see p1634157647114600-slack-C5ALGJYEP) and pushed the spec in our internal spec repo [here](https://github.com/wordpress-mobile/cocoapods-specs/blob/b999b32dd16d318224a440bf2b4f33d6906d49e7/MobilePayKit/0.0.2/MobilePayKit.podspec)
 - The corresponding Buildkite passing job is [this one](https://buildkite.com/wordpress-mobile/mobilepaykit-ios/builds/14).

### 🧹  Once this was validated, I cleaned things up by:

 - Deleting the `0.0.2` tag from this repo (it was on a commit of the branch anyway while tags should be on `develop` for pod repos consistency)
 - Reverting adf8ca8, 9a48b7a and 00b6a9f to re-enable the `pod trunk push` to the CDN and point back to `0.0.1`
 - Deleting the `MobilePayKit.podspec` file from our spec repo (via https://github.com/wordpress-mobile/cocoapods-specs/commit/04eb422546c70a73f18af6e43b24214cd746fa26) to "unpublish" this fake `0.0.2`

_Note that given I disabled `pod trunk push` during my tests, the fake `0.0.2` was never pushed to the CDN (that also means that the push to CDN step hasn't been tested in this PR, but it shouldn't cause any trouble 🤞 since we already use this same config for other pod repos)_

The current state of this PR is now after this validation and cleanup, which means the PR should now in a state that is ready for review 🎉 

### 🚨 To fix before merge

We still need to wait for a new release of the `bash-cache-buildkite-plugin` after https://github.com/Automattic/bash-cache-buildkite-plugin/pull/8 lands to update line 4 of the the `pipeline.yml` file and point to the stable version of the plugin.
\cc @jkmassel 

### ℹ️  A note about the `0.0.1` tag

The tag `0.0.1` – and thus the corresponding version in CocoaPods CDN and our spec repo – won't be able to be pushed in its current state, because it points to a commit which generates a `xcodebuild` warning during compilation ("unused variable"), while we don't allow warnings for the pods we publish (in other words, we don't use `--allow-warnings` when calling `pod <repo|trunk> push`). \cc @momo-ozawa fyi.

That warning has been fixed in this PR – see c2fa6ec – so once this is merged this won't be a problem anymore, but the `0.0.1` doesn't contain this fix.

We will thus need to:
 - Either delete the tag `0.0.1`, assuming nobody uses it yet. That's a big assumption, which only the people working on this repo and pod might know the answer to for sure, so I'd be very wary about doing that, except if everyone in the project is on board. In general, I'm not a fan of deleting tags, but this project is so early in its steps, that… 🤷 
 - Or keep `0.0.1` and accept that the tag will never be published to CocoaPods, and do a new tag (`0.0.2` or `0.1.0` for example) – and update the `.podspec` accordingly, of course – once we're ready to publish our first version to CocoaPods. I'd recommend this approach instead.

File: .buildkite/pipeline.yml
Comment (AliSoftware) [inline .buildkite/pipeline.yml:4]: 🚨  We need to update this reference once https://github.com/Automattic/bash-cache-buildkite-plugin/pull/8 lands and we made a new version of this plugin

**Comment by momo-ozawa** [comment]
@AliSoftware 👋 

Thank you for tackling this task!

> Or keep 0.0.1 and accept that the tag will never be published to CocoaPods, and do a new tag (0.0.2 or 0.1.0 for example) – and update the .podspec accordingly, of course – once we're ready to publish our first version to CocoaPods. I'd recommend this approach instead.

This sounds good to me. Tagging @leandroalonso @emilylaguna for visibility.

**Comment by emilylaguna** [comment]
Sounds good to me!

File: .buildkite/pipeline.yml
Comment (mokagio) [inline .buildkite/pipeline.yml:6]: Any particular reason for using 12.5.1 as opposed to 13.0 (`xcode-13`)?

File: MobilePayKit.podspec
Comment (mokagio) [inline MobilePayKit.podspec:7]: Nitpick 

```suggestion
  s.summary       = "Automattic's iOS and macOS Mobile Payments library"
```

Comment (mokagio) [reply MobilePayKit.podspec:7]: Or
```suggestion
  s.summary       = "Automattic's iOS and macOS mobile payments library'
```

**Review by mokagio — COMMENTED** [review]
Looking good so far. Holding on ✅ till the new `bash-cache` version is available.

I checked out locally and run the `ios test` lane 👍

File: .buildkite/pipeline.yml
Comment (AliSoftware) [reply .buildkite/pipeline.yml:6]: Not a really strong one no. Mostly used this because iinmc most of our other pods and developers are still using Xcode 12.5 too (until we proof-test them under Xcode 13 and validate that we're ready to migrate?)

File: MobilePayKit.podspec
Comment (AliSoftware) [reply MobilePayKit.podspec:7]: I'm not sure that means the same thing.

Your suggestion means that "this is an (all-purpose) mobile payment library (that anybody can use in their Open-Source, unrelated project), and it happens that this library is authored by Automattic"

While my wording is "this is a mobile payment library dedicated for Automattic apps (and possibly relying on its infrastructure)"

I'm not sure which one is more correct? \cc @momo-ozawa

Comment (momo-ozawa) [reply MobilePayKit.podspec:7]: >  "this is a mobile payment library dedicated for Automattic apps (and possibly relying on its infrastructure)"

This is the meaning we'd want to go for, so I'm casting a vote for `iOS/macOS Mobile Payments library for Automattic` 👍

Comment (mokagio) [reply MobilePayKit.podspec:7]: Cool. Thanks for the clarification 😄

File: .bundle/config
Comment (jkmassel) [inline .bundle/config:4]: This seems like an odd thing to hardcode into the project (what if I run `bundle` on a 4-core machine?)

File: .buildkite/pipeline.yml
Comment (jkmassel) [reply .buildkite/pipeline.yml:6]: I'd say we could do the 13 upgrade in a separate PR

File: .gitignore
Comment (jkmassel) [inline .gitignore:2]: nice!

File: MobilePayKit.podspec
Comment (jkmassel) [inline MobilePayKit.podspec:5]: Well this seems like a regression 😜

**Review by jkmassel — CHANGES_REQUESTED** [review]
This is generally good, but I'm curious – why is all the macOS stuff disabled?

File: .bundle/config
Comment (AliSoftware) [reply .bundle/config:4]: This is related to p91TBi-6it-p2. This particular key is inserted by `bundle install` itself on every run (i.e. I didn't add it manually myself)

File: MobilePayKit.podspec
Comment (AliSoftware) [reply MobilePayKit.podspec:5]: It's not.

The `1.0.0` tag never existed so far, and the `podspec` was created by the team with this value in there… even before the pod was pushed anywhere (and in fact it's still not pushed to trunk nor our private spec repo, since this PR was needed for this to happen), and while the [only tag that has been created (manually) by the team was `0.0.1`](https://github.com/Automattic/MobilePay-Apple/tags). So this is just fixing the initial `podspec` they wrote to match the tag they created.

**Comment by AliSoftware** [comment]
> I'd say we could do the 13 upgrade in a separate PR

Yep this PR was opened way before we even started working on the Xcode 13 migration anyway 😅  so I agree we should merge this first and consider migrating to Xcode 13 separately.

> This is generally good, but I'm curious – why is all the macOS stuff disabled?

Simply because the current UI tests the team has written for this pod don't support running on macOS (this is mainly an architecture restriction setting).

I figured that if they wanted to write macOS-compatible tests I'd leave those ready to be un-commented, but the goal was to address that request ASAP(†) rather than spend time working on the not-yet-existing/compatible macOS unit tests

_(†) well clearly with the review delay that premise has failed 😂  but still, given the time this has been waiting, I think it's still better to merge this first, and consider adding macOS Unit Tests to the project in a separate PR._

File: MobilePayKit.podspec
Comment (jkmassel) [reply MobilePayKit.podspec:5]: Makes sense – but we're still not at `1.0.0` yet? This'll be an in-development dependency for a while yet?

**Review by jkmassel — APPROVED** [review]
>I figured that if they wanted to write macOS-compatible tests I'd leave those ready to be un-commented, but the goal was to address that request ASAP(†) rather than spend time working on the not-yet-existing/compatible macOS unit tests

Makes sense!

It's a :shipit: from me

File: MobilePayKit.podspec
Comment (AliSoftware) [reply MobilePayKit.podspec:5]: Yep I think it's still a `0.0.1` version as it's being worked on by the team. @momo-ozawa will probably confirm.

According to SemVer, a `0.x.y` version is totally valid to be pushed to CocoaPods anyway, and just signals that "the API of this library can change at any point because it's in its early stage of development so we're not commited to have the API being stable for a major version just yet"

---

The next step once this PR is merged, is for the team to continue working on the codebase until they consider that  the API of the library is stable enough to make a first major `1.0.0` version, or even an intermediate `0.y.z` version beforehand,:only then will they update the `podspec` with the next version number they deem appropriate only then.

But we obviously need to merge this PR first _before_ doing that and considering creating a new version and tag, which is why I used `0.0.1` here to match the current tag and will let the team update that value only when they're ready to make a new version and decide which version number that next one should be.

---

PS: Note that merging this PR will _not_ publish the already-existing `0.0.1` tag – which as I pointed out in [this previous comment](https://github.com/Automattic/MobilePay-Apple/pull/19#issuecomment-942750596), won't be able to be pushed to trunk anyway, due to compilation warnings.

**Comment by AliSoftware** [comment]
Next steps \cc @momo-ozawa @leandroalonso @emilylaguna and team:

## Xcode 13 support

Since this PR was opened a couple of days ago, it was still pointing to Xcode 12.5
We've now migrated our CI to Xcode 13, so a follow-up PR to update the CI image to Xcode 13 would thus make sense. 

Don't hesitate to open such a PR once you've validated your codebase is ready to migrate to Xcode 13.

The only change that should be needed for this on the CI config side is to update line 6 of `pipeline.yml` (the `IMAGE_ID` field), but please ensure that compiling your codebase with Xcode 13 doesn't generate new warnings first (that would prevent the pod from being published) and address them if needed.

## Making Unit Tests run on macOS

The Unit Tests are only run for iOS on CI, because the Xcode target `MobilePayKitTests` is configured for iOS and does not support macOS architectures.

I'm not entirely sure it is possible to configure a single test target to be able to be run on both platforms – especially providing a dynamic "Base SDK" build setting depending on the target architecture? So maybe you'll need to duplicate the test target, containing the same test files and settings except for "Supported Platform", "Base SDK" and "Target Deployment"?
\cc @mokagio as our test guru, if you have any insight on how to make a single test target be run on different Base SDK without duplicating it?

Once you have Unit Tests working on macOS, you will be able to uncomment the corresponding lines [in the `pipeline.yml`](https://github.com/Automattic/MobilePay-Apple/pull/19/files#diff-a3991ebf1475eb82acab13946ffc1eca02b43917f6d5bec3898f45c8b0b9bd53R20-R25) and [`Fastfile`](https://github.com/Automattic/MobilePay-Apple/pull/19/files#diff-dff4b99d4e651ab9d7597876ec8dbe92aa934e42c0fb55f4aadd6c106fc96688R20-R30) to enable those to run on CI.

## Publishing new versions

You probably already know this, but just in case and because it's always good to be explicit and have a nice checklist:

When you're ready to publish a new version of this pod, just:
 - Create a `release/x.y.z` branch, update the `.podspec` file's `s.version` field, and make a PR to merge it
 - Wait for the PR to go green: it should in particular confirm that the pod doesn't have any warning and is ready to be published
 - Once the PR is merged, create a GitHub release with the same version name, making it create a namesake git tag once it's published.
 - The creation of the tag will be what triggers the publication of the pod to both CocoaPods trunk and our internal private spec repo (which should make it available sooner than having to wait for the CDN to update and propagate it)
 - You can track the progress of the publication of new tags [on Buildkite here](https://buildkite.com/wordpress-mobile/mobilepaykit-ios). Once the Buildkite job has succeeded, you can update your client apps to point to the new version.

Reminder that the existing `0.0.1` tag won't be published to CocoaPods (due to warnings present on that tag).

File: .buildkite/pipeline.yml
Comment (AliSoftware) [reply .buildkite/pipeline.yml:4]: And of course I forgot to do this before merging… Fixing this in #21

**Comment by mokagio** [comment]
I haven't had a chance to look at the test setup for this yet, but I'm definitely intrigued. I know there are various open source projects that build for different platforms, we might want to start there for inspiration on the setup.

**Pull request merged**
